### PR TITLE
Feat/military affiliation records

### PR DESCRIPTION
**Pull request by landonshumway-ia**
We need to support licensees uploading documentation as evidence for their military affiliation. We will be recording the military affiliation upload history as records in the provider DynamoDB table which will be used by the system to determine a user's current military affiliation status. As part of this, we've added a new POST endpoint which allows the user to specify a filename. When called, the API generates a S3 pre-sign URL (active for 10 minutes) which the front end will then use to upload that file into S3 directly. Once the file is successfully uploaded, S3 will trigger a lambda function to complete the initialization of the military affiliation record for that provider so they will be eligible for any military discounts when purchasing privileges. 

When a user wants to end their military affiliation, the client will call a PATCH endpoint that will set their military status to inactive.

### Requirements List
- No new requirements for deployment. All changes are backwards compatible.

### Description List
- Add S3 bucket, which will be used to store licensee documents that they upload into the system.
- Add event trigger to handle upload events to this bucket
- Add POST api endpoint for creating military affiliation records
- Add PATCH api endpoint for deactivating military affiliation status.
- Add Smoke tests which can be run locally against a sandbox environment to verify the military document upload flow.
- Update Privilege purchase endpoint to check user's military affiliation status when determining if they qualify for any military discounts.
- Update provider data endpoint to return list of military affiliation records as part of their profile information

### Testing List
- `yarn test:unit:all` should run without errors or warnings
- `yarn serve` should run without errors or warnings
- `yarn build` should run without errors or warnings
- For API configuration changes: CDK tests added/updated in `backend/compact-connect/tests/unit/test_api.py`
- Local smoke tests
- Code review

Closes #[327](https://github.com/csg-org/CompactConnect/issues/327)

### CODE REVIEW

**Comment by landonshumway-ia** [comment]
@jlkravitz - Given that @jusdino is out for the week, I'm putting this up for CSG review. Feel free to postpone on this review if you would prefer to wait until @jusdino gets back. Thanks.

**Comment by jlkravitz** [comment]
Happy to take a pass now. I assume @jusdino will review when he's back?

- [ ] review the pull request to get oriented
	- [ ] read the description of the pull request, which should summarize the changes made
	- [ ] read through every task on the Scrum board that's encompassed by this pull request
	- [ ] read the description of the commits that comprise the pull request
- [ ] skim all new code, in the context of existing code, looking for problems (knowing that the vast majority of new code will be covered by tests)
- [ ] review all tests
	- [ ] methodically review all new tests for correctness, quality of naming
	- [ ] look at code coverage of tests
	- [ ] determine what code isn’t tested, review that rigorously
- [ ] review documentation to ensure that it matches changes
- [ ] provide comments on the pull request on GitHub, as necessary
	- [ ] for comments that are specific to a particular line of code, comment on those specific lines
	- [ ] for comments that are more general, attach the comment to a random line in `README.md` (as opposed to commenting on the pull request itself), to be able to use GitHub's ability to thread discussions on those comments
- [ ] run a security audit of dependencies (e.g. `npm audit` and `pip audit`) to ensure that there are no vulnerabilities that will be deployed to production (as opposed to vulnerabilities that only have an impact on the development environment)

File: backend/compact-connect/lambdas/provider-data-v1/tests/function/test_data_model/test_client.py
Comment (jlkravitz) [inline backend/compact-connect/lambdas/provider-data-v1/tests/function/test_data_model/test_client.py:308]: Is this something we can assert?

File: backend/compact-connect/lambdas/provider-data-v1/tests/function/test_data_model/test_provider_transformations.py
Comment (jlkravitz) [inline backend/compact-connect/lambdas/provider-data-v1/tests/function/test_data_model/test_provider_transformations.py:133]: Is this comment accurate?

File: backend/compact-connect/lambdas/common-python/tests/unit/test_data_model/test_data_client.py
Comment (jlkravitz) [inline backend/compact-connect/lambdas/common-python/tests/unit/test_data_model/test_data_client.py:11]: Conceptual question: Which records are you intending to delete if there's an exception?

File: backend/compact-connect/tests/smoke/military_affiliation_smoke_tests.py
Comment (jlkravitz) [inline backend/compact-connect/tests/smoke/military_affiliation_smoke_tests.py:15]: How does one get this token if they want to run these smoke tests? Can we spell this out a bit more (e.g., the recommended workflow for getting a token)?

File: backend/compact-connect/tests/smoke/military_affiliation_smoke_tests.py
Comment (jlkravitz) [inline backend/compact-connect/tests/smoke/military_affiliation_smoke_tests.py:58]: nit: I'm used to this format of docstring only being used for function, class, or module comments. Would it be more appropriate to use `#` like you do at the start of this function?

File: backend/compact-connect/tests/smoke/military_affiliation_smoke_tests.py
Comment (jlkravitz) [inline backend/compact-connect/tests/smoke/military_affiliation_smoke_tests.py:92]: Why this vs. just `print`?

**Review by jlkravitz — COMMENTED** [review]
Generally looks great! A few comments/questions. Thanks for documenting the changes thoroughly- made it easy to follow.

File: backend/compact-connect/tests/smoke/military_affiliation_smoke_tests.py
Comment (landonshumway-ia) [reply backend/compact-connect/tests/smoke/military_affiliation_smoke_tests.py:92]: The Ruff linter flags us for use of `print` statements in code, so this is the workaround

File: backend/compact-connect/lambdas/common-python/cc_common/data_model/client.py
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/common-python/cc_common/data_model/client.py:361]: So @jusdino  and I walked through this scenario, and determined if a user attempted to upload new documentation, and that documentation upload failed, we will want to let them know. By setting the previous record as inactive, we are forcing the system to notify the user that they need to reupload their documentation before they can proceed with purchasing privileges.

File: backend/compact-connect/lambdas/common-python/cc_common/data_model/client.py
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/common-python/cc_common/data_model/client.py:370]: Yeah that could work, it just means we would be making two API batch calls to Dynamo, rather than a single batch call. So slightly less performant, but not by much.

File: backend/compact-connect/lambdas/provider-data-v1/tests/function/test_data_model/test_provider_transformations.py
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/provider-data-v1/tests/function/test_data_model/test_provider_transformations.py:133]: yes, because the test only calls 'create_military_affiliation' and not the s3 event trigger, the military affiliation record will be in an 'initializing' state.

File: backend/compact-connect/tests/smoke/military_affiliation_smoke_tests.py
Comment (landonshumway-ia) [reply backend/compact-connect/tests/smoke/military_affiliation_smoke_tests.py:58]: In this case, I was intending to have a multi-line comment, not a docstring. Looking it up I see the convention for multi line strings is single quotes. I'll update this to reflect that.

File: backend/compact-connect/tests/smoke/military_affiliation_smoke_tests.py
Comment (landonshumway-ia) [reply backend/compact-connect/tests/smoke/military_affiliation_smoke_tests.py:15]: I've added a comment walking through the simplest steps to get the id token.

File: backend/compact-connect/stacks/api_stack/v1_api/api_model.py
Comment (landonshumway-ia) [reply backend/compact-connect/stacks/api_stack/v1_api/api_model.py:217]: That was an initial draft that I used so copilot would generate all the json schema boilerplate for me. I'll remove it now since it's out of date.

File: backend/compact-connect/lambdas/provider-data-v1/tests/function/test_data_model/test_client.py
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/provider-data-v1/tests/function/test_data_model/test_client.py:295]: good catch

File: backend/compact-connect/lambdas/common-python/cc_common/data_model/schema/common.py
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/common-python/cc_common/data_model/schema/common.py]: This is actually not needed, it is leftover generated copilot code from when I was trying to lookup military enum values for logic I ended up not using. I've removed the redundant check.

File: backend/compact-connect/lambdas/common-python/cc_common/data_model/client.py
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/common-python/cc_common/data_model/client.py:370]: I could go either way, let me know if you have a strong preference and I'll update as needed.

**Comment by landonshumway-ia** [comment]
@jlkravitz Once this is approved, we could have @jusdino do a post-merge review when he gets back next week so this branch doesn't become stale, or we could wait to merge until he performs a review. I'll leave that up to your call.

File: backend/compact-connect/lambdas/common-python/cc_common/data_model/client.py
Comment (jlkravitz) [reply backend/compact-connect/lambdas/common-python/cc_common/data_model/client.py:361]: Got it. That makes sense!

File: backend/compact-connect/lambdas/common-python/cc_common/data_model/client.py
Comment (jlkravitz) [reply backend/compact-connect/lambdas/common-python/cc_common/data_model/client.py:370]: I vote the two API call approach - perhaps slightly less performant, but more readable imo. Perhaps add a brief comment above the second call like `// These could be consolidated into a single batch call if performance becomes an issue`

File: backend/compact-connect/tests/smoke/military_affiliation_smoke_tests.py
Comment (jlkravitz) [reply backend/compact-connect/tests/smoke/military_affiliation_smoke_tests.py]: Good to know!

File: backend/compact-connect/tests/smoke/military_affiliation_smoke_tests.py
Comment (jlkravitz) [reply backend/compact-connect/tests/smoke/military_affiliation_smoke_tests.py:93]: Ah strange. Can we use print, but suppress the linter on those lines?

I'm fine with keeping this as-is, too, but can you add a comment noting why you're not using print here? Can be brief, but it does stand out without some context.

File: backend/compact-connect/lambdas/common-python/tests/unit/test_data_model/test_data_client.py
Comment (jlkravitz) [reply backend/compact-connect/lambdas/common-python/tests/unit/test_data_model/test_data_client.py:11]: Got it- very helpful!

File: backend/compact-connect/lambdas/provider-data-v1/tests/function/test_data_model/test_provider_transformations.py
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/provider-data-v1/tests/function/test_data_model/test_provider_transformations.py:133]: ah sorry, I was looking at the lines about this, no that comment looks outdated, I'll remove it.

File: backend/compact-connect/tests/smoke/military_affiliation_smoke_tests.py
Comment (landonshumway-ia) [reply backend/compact-connect/tests/smoke/military_affiliation_smoke_tests.py]: I've suppressed the linter and replaced the print statements to improve readability

**Review by jlkravitz — APPROVED** [review]
@isabeleliassen Good to merge. @jusdino has not reviewed this PR since he is OOO, but per @landonshumway-ia's comment, I think it makes most sense to merge this now and have @jusdino do a post-merge review.

**Review by isabeleliassen — APPROVED** [review]

**Pull request merged**
